wz_strategic_plugin/src/wz_strategic_plugin.cpp

```diff
@@ -0,0 +1,508 @@
+
+  // Identify the lanelets which will be crossed by approach maneuvers lane follow maneuver
+  std::vector<lanelet::ConstLanelet> crossed_lanelets =
+      getLaneletsBetweenWithException(current_state.downtrack, intersection_end_downtrack, true, false);
+
+  // Compose intersection transit maneuver
+  resp.new_plan.maneuvers.push_back(composeIntersectionTransitMessage(
+      current_state.downtrack, intersection_end_downtrack, intersection_speed_limit, intersection_speed_limit,
+      current_state.stamp, intersection_exit_time, crossed_lanelets.front().id(), crossed_lanelets.back().id()));
+}
+
+bool WzStrategicPlugin::planManeuverCb(cav_srvs::PlanManeuversRequest& req, cav_srvs::PlanManeuversResponse& resp)
+{
+  if (!wm_->getRoute())
+  {
+    ROS_ERROR_STREAM("Could not plan maneuvers as route was not available");
+    return true;
+  }
+
+  ROS_DEBUG("Finding car information");
+
+  // Extract vehicle data from request
+  VehicleState current_state = extractInitialState(req);
+
+  // Get current traffic light information
+  ROS_DEBUG("\n\nFinding traffic_light information");
+
+  auto traffic_list = wm_->getLightsAlongRoute({ req.veh_x, req.veh_y });
+
+  if (traffic_list.empty())  // If there were no lights found on the route then this plugin is not applicable
+  {
+    ROS_DEBUG("No lights found along route. Returning maneuver plan unchanged");
+    // Return empty maneuver plan unchanged
+    return true;
+  }
+
+  TransitState prev_state;
+
+  do
+  {
+    // Clear previous maneuvers planned by this plugin as guard against state change since each state generates an independant set of maneuvers
+    resp.new_plan =  cav_msgs::ManeuverPlan();
+
+    prev_state = transition_table_.getState();  // Cache previous state to check if state has changed after 1 iteration
+
+    switch (transition_table_.getState())
+    {
+      case TransitState::UNAVAILABLE:
+        planWhenUNAVAILABLE(req, resp, current_state, traffic_list);
+        break;
+
+      case TransitState::APPROACHING:
+        planWhenAPPROACHING(req, resp, current_state, traffic_list);
+        break;
+
+      case TransitState::WAITING:
+        planWhenWAITING(req, resp, current_state, traffic_list);
+        break;
+
+      case TransitState::DEPARTING:
+        // Reset intersection state since we are no longer in an intersection
+        if (!intersection_end_downtrack_ || !intersection_speed_)
+        {
+          throw std::invalid_argument("State is DEPARTING but the intersection variables are not available");
+        }
+        planWhenDEPARTING(req, resp, current_state, intersection_end_downtrack_.get(),
+                          intersection_speed_.get());
+        break;
+
+      default:
+        throw std::invalid_argument("WorkZone Strategic Plugin entered unknown state");
+        break;
+    }
+
+  } while (transition_table_.getState() != prev_state);  // If the state has changed then continue looping
```

Comment:
Is the state going to change multiple times within one planManeuver call? 
If we are within range of the intersection (lets say 50m away) will we ever exit this loop? 
Maybe i'm thinking of this wrong but prev state and current_state will be always be the same unless we are really close to moving from one state to another.

Reply:
The state can change multiple times on one loop if the vehicle has satisfied multiple events within the 1s timing period of maneuver planning. For example:
- leaving an intersection and immediately approaching another intersection that is within stopping range

I elected to allow all the states to fully transition in this way instead of having to wait a whole 1s for the next plan before allowing a new state transition. 

I'm not sure I see where the risk of not exiting the loop comes from, can you elaborate?